source/src/localisation/cs_TrackableParser.cpp

```diff
@@ -124,136 +223,371 @@ void TrackableParser::handleScannedDevice(scanned_device_t* device) {
+	LOGTrackableParserDebug("handle get filter summaries:");
+	for (auto& trackingFilter : _parsingFilters) {
+		if (trackingFilter == nullptr) {
+			LOGTrackableParserDebug("filter: nullptr");
+			continue;
+		}
+		LOGTrackableParserDebug("filter: %u", trackingFilter->runtimedata.filterId);
+	}
+
+	// stack allocate a buffer summaries object fitting at most max summaries:
+	auto requiredBuffSize = sizeof(trackable_parser_cmd_get_filter_summaries_ret_t)
+							+ sizeof(tracking_filter_summary_t) * _parsingFiltersCount;
+
+	if (result.buf.len < requiredBuffSize) {
+		result.returnCode = ERR_BUFFER_TOO_SMALL;
+		return;
+	}
+
+	// placement new constructs the object in the buff and now has enough space for the summaries.
+	auto retvalptr  = new (result.buf.data) trackable_parser_cmd_get_filter_summaries_ret_t;
+	result.dataSize = requiredBuffSize;
+
+	retvalptr->masterCrc     = _masterHash;
+	retvalptr->masterVersion = _masterVersion;
+	retvalptr->freeSpace     = getTotalHeapAllocatedSize();
+
+	for (size_t i = 0; i < _parsingFiltersCount; i++) {
+		retvalptr->summaries[i].id      = _parsingFilters[i]->runtimedata.filterId;
+		retvalptr->summaries[i].crc     = _parsingFilters[i]->runtimedata.crc;
+		retvalptr->summaries[i].flags   = _parsingFilters[i]->metadata.flags;
+		retvalptr->summaries[i].version = _parsingFilters[i]->metadata.version;
+	}
+
+	result.returnCode = ERR_SUCCESS;
+}
+
+// -------------------------------------------------------------
+// ---------------------- Utility functions --------------------
+// -------------------------------------------------------------
+void TrackableParser::startProgress() {
+	filterModificationInProgress = true;
+	_masterVersion               = 0;
+}
+
+uint16_t TrackableParser::masterCrc() {
+	uint16_t masterCrc = 0;
+	uint16_t* prevCrc  = nullptr;
+
+	for (auto trackingFilter : _parsingFilters) {
+		if (trackingFilter == nullptr) {
+			break;
+		}
+
+		LOGTrackableParserDebug("filter crc: %x", trackingFilter->runtimedata.crc);
+
+		// appends/applies this filters' crc onto master crc
+		masterCrc =
+				crc16(reinterpret_cast<const uint8_t*>(&trackingFilter->runtimedata.crc),
+					  sizeof(trackingFilter->runtimedata.crc),
+					  prevCrc);
+
+		prevCrc = &masterCrc;  // first iteration will have prevCrc == nullptr, all subsequent ones point to masterCrc.
+	}
+
+	LOGTrackableParserDebug("master crc: %x", masterCrc);
+
+	return masterCrc;
+}
+
+bool TrackableParser::checkFilterSizeConsistency() {
+	bool consistencyChecksFailed = false;
+
+	size_t index = 0;
+	while (index < _parsingFiltersCount) {
+		auto trackingFilter = _parsingFilters[index];
+
+		if (trackingFilter == nullptr) {
+			// early return: last filter has been handled.
+			break;
+		}
+
+		if (trackingFilter->runtimedata.crc == 0) {
+			// filter changed since commit.
+
+			size_t trackingFilterSize = getTotalSize(*trackingFilter);
+			auto cuckoo               = CuckooFilter(trackingFilter->filterdata);
```

Comment:
This can't cause any problems if the filter data is invalid?

Reply:
I can maybe move the crc check in commitFilterCommands up a tiny bit so that this check is less likely to hit exceptions